ambry-mysql/src/main/java/com/github/ambry/mysql/MySqlDataAccessor.java

```diff
@@ -269,7 +269,9 @@ public static boolean isCredentialError(SQLException e) {
    * @param operationType type of mysql operation
    */
   public void onException(SQLException e, OperationType operationType) {
-    if (e instanceof SQLTransientConnectionException) {
+    // Close connection for all non transient sql exceptions.
+    if (!(e instanceof SQLNonTransientException || (e instanceof BatchUpdateException
+        && e.getCause() instanceof SQLNonTransientException))) {
       if (operationType == OperationType.Write) {
```

Comment:
Should we move the metrics code outside the exception type check?  Seems like we should record the errors even if it was a non-transient exception.